sourced/ml/transformers/indexer.py

```diff
@@ -0,0 +1,60 @@
+import typing
+from pyspark.rdd import RDD
+from pyspark import Row
+
+from sourced.ml.transformers import Transformer
+
+
+class Indexer(Transformer):
+    """
+    Transformer that creates index mapping for one column in spark rdd.
+    It map all values in the column to its index.
+    One can get python dictionary with mapping from column2id property and unique values list from
+    column_values.
+    """
+    def __init__(self, column: typing.Union[int, str], **kwargs):
+        """
+        :param column: column index or its name in pyspark.RDD for indexing.
+        :param kwargs:
+        """
+        super().__init__(**kwargs)
+        self.column = column
+        self._column2id = None
+        self._column_values = None
+
+    @property
+    def column2id(self):
+        if self._column2id is None:
+            raise RuntimeError("column2id value not available. Run Indexer first.")
+        return self._column2id
+
+    @property
+    def column_values(self):
+        if self._column_values is None:
+            raise RuntimeError("column_values value not available. Run Indexer first.")
+        return self._column_values
+
+    def __call__(self, rdd: RDD):
+        column_id = self.column
+        if isinstance(column_id, str):
+            column = rdd.map(lambda x: (getattr(x, column_id),))
+        else:
+            column = rdd.map(lambda x: (x[column_id],))
+
+        self._column_values = column \
+            .distinct() \
+            .map(lambda x: x[0]) \
+            .collect()
+
+        column2id = {d: i for i, d in enumerate(self._column_values)}
+        self._column2id = column2id
+
+        def index_column(row):
```

Comment:
This is suboptimal. PySpark will attempt to pickle this function. This is a scoped function which has `self` in the closure. This means `self` will be pickled too and subsequently the mappings. We only need `column2id` and do not need the second one.

Fix: `functools.partial` on a `@staticmethod` or a class.

Reply:
Ok, it seems to be true, but if I use `self` inside this function. And in this case, pyspark fails to pickle it for some reason. So I do such things as `column_id = self.column` and so on. It seems that spark optimizes something and do not try to pickle self, if I don't use it inside. 

I did the following to check my words: add `__getstate__` to this class:
```python
    def __getstate__(*args, **kwargs):
        print("\nGEEEEEET STAAAAAATE\n")
```

Nothing was broke and I did not see the output. I will try to find some explanation in the documentation about such behavior. I prefer to leave it as is

Reply:
Such things are implementation specific. It may work, and may not in other interpreters, Spark versions, etc. Please be explicit. It's not hard to refactor.

Reply:
it is ok to do as I did (http://spark.apache.org/docs/latest/rdd-programming-guide.html#passing-functions-to-spark). I Discuss it with @vmarkovtsev and decide to add comment.